fix: Open the correct menu for folder nav items

dgrothaus-sw opened the pull request:
Fixed opening the correct menu when hovering a `folder` type navigation item. Formerly the menu of another nav item was opened when hovering a `folder` nav item.
Added an option to omit adding the click event listener for nav items that don't have a hypertext reference

This is a forward port of https://github.com/shopware/shopware/pull/7417
Fixes https://github.com/shopware/shopware/pull/7417

dgrothaus-sw committed:
fix: Open the correct menu for folder nav items

Fixed opening the correct menu when hovering a `folder` type navigation
item. Formerly the menu of another nav item was opened when hovering a
`folder` nav item.
Added an option to omit adding the click event listener for nav items
that don't have a hypertext reference

This is a forward port of https://github.com/shopware/shopware/pull/7417
Fixes https://github.com/shopware/shopware/pull/7417

dgrothaus-sw commented:
Please tell me your thoughts about moving the closing `</li>` outside of the control structure that opens the element. I'm not entirely sure if this is the best solution, but the other solution would be to have the exact same lines for the child menu twice in the `{% if %}`and the `{% else %}`

tobiasberge commented on src/Storefront/Resources/app/storefront/src/plugin/navbar/navbar.plugin.js:
```diff
@@ -31,7 +31,9 @@ export default class NavbarPlugin extends Plugin {
             el.addEventListener(openEvent, this._toggleNavbar.bind(this, el));
             el.addEventListener(closeEvent, this._toggleNavbar.bind(this, el));
-            el.addEventListener(clickEvent, this._navigateToLinkOnClick.bind(this, el));
+            if (el.dataset?.flyoutMenuWithoutTarget !== 'yes') {
```
Instead of introducing this new `data-flyout-menu-without-target="yes"` couldn't we check if the given element has an href? Inside `_navigateToLinkOnClick` the href is directly used. href should be `null` when the attribute cannot be found.


```suggestion
            if (el.getAttribute('href') !== null) {
```

dgrothaus-sw replied:
Good point. Will change accordingly :slightly_smiling_face:

dgrothaus-sw committed:
fix: Check for href instead of introducing flag

Don't introduce a dedicated flag for a single purpose but check for the
existance of an href attribute.

tobiasberge commented:
> Please tell me your thoughts about moving the closing `</li>` outside of the control structure that opens the element. I'm not entirely sure if this is the best solution, but the other solution would be to have the exact same lines for the child menu twice in the `{% if %}`and the `{% else %}`

@dgrothaus-sw I would advocate for duplicating the include of the children. And not having only an opening `<li>` within the `{% if %}`. Ideally the opening and closing twig and html nodes should be coherent and not close itself in some other "scope". Even if the current control structure would not allow a wrong rendering in the final HTML.

But this makes the template harder to scan and over time html errors can slip through the cracks because it is hard to see where which element is closing. E.g. when future-somebody is doing some changes further down in the template. Opening and closing the `<li>` in the same twig if is cleaner imo, even if we have some duplicate lines. Sometimes also the IDEs are showing errors because they don't understand the structure.

And ludtwig would also throw an error there: https://github.com/shopware/shopware/pull/7393

dgrothaus-sw committed:
dix: Be less DRY to satisfy ludtwig

dgrothaus-sw commented:
> @dgrothaus-sw I would advocate for duplicating the include of the children…

Done.

dgrothaus-sw committed:
Merge branch 'trunk' into next-40641/nav-menu-for-folder-nav-items

tobiasberge commented on src/Storefront/Resources/views/storefront/layout/navbar/navbar.html.twig:
```diff
@@ -33,12 +33,23 @@
                                 {% if category.type == 'folder' %}
-                                    <li class="nav-item main-navigation-link nav-item-{{ id }} nav-link root nav-item-{{ id }}-link"
+                                    <li class="nav-item main-navigation-link nav-item-{{ id }} nav-link root nav-item-{{ id }}-link{% if hasChildren %} dropdown-toggle{% endif %}"
+                                        data-bs-toggle="dropdown"
```
I think `data-bs-toggle="dropdown"` should also be inside `{% if hasChildren %}`

dgrothaus-sw committed:
fix: use dropdown toggle only for child elements
